File: src/services/AccountService.cpp
#include "services/AccountService.h"

#include <cmath>

#include "core/Utils.h"

namespace bms {

Account* AccountService::openAccount(int customerId, AccountType type,
                                     long long openingDeposit,
                                     std::string& err) {
    if (db_.findCustomerById(customerId) == nullptr) {
        err = "No such customer.";
        return nullptr;
    }
    if (openingDeposit < 0) {
        err = "Opening deposit cannot be negative.";
        return nullptr;
    }

    Account acc;
    acc.accountNumber = db_.nextAccountNumber();
    acc.customerId = customerId;
    acc.type = type;
    acc.balancePaise = openingDeposit;
    acc.interestRate = type == AccountType::SAVINGS ? 3.5 : 0.0;
    acc.status = AccountStatus::ACTIVE;
    acc.createdAt = Utils::now();
    acc.lastInterestApplied = Utils::today();
    db_.accounts().push_back(acc);

    if (openingDeposit > 0) {
        Transaction tx;
        tx.id = db_.nextTransactionId();
        tx.accountNumber = acc.accountNumber;
        tx.type = TxType::DEPOSIT;
        tx.amountPaise = openingDeposit;
        tx.balanceAfterPaise = openingDeposit;
        tx.counterparty = "Bank";
        tx.timestamp = Utils::now();
        tx.remark = "Account opening deposit";
        db_.addTransaction(tx);
    }

    db_.saveAll();
    return &db_.accounts().back();
}

bool AccountService::closeAccount(const std::string& accountNumber,
                                  std::string& err) {
    Account* acc = db_.findAccountByNumber(accountNumber);
    if (!acc) {
        err = "Account not found.";
        return false;
    }
    if (acc->status != AccountStatus::ACTIVE) {
        err = "Account is already closed.";
        return false;
    }
    if (acc->balancePaise != 0) {
        err = "Balance must be zero before closing. Withdraw or transfer first.";
        return false;
    }
    for (const auto& loan : db_.loans()) {
        if (loan.accountNumber == accountNumber &&
            (loan.status == LoanStatus::ACTIVE || loan.status == LoanStatus::PENDING)) {
            err = "Account has an open loan linked to it.";
            return false;
        }
    }

    acc->status = AccountStatus::CLOSED;
    db_.saveAll();
    return true;
}

std::vector<const Account*> AccountService::accountsOfCustomer(int customerId) const {
    std::vector<const Account*> out;
    for (const auto& a : db_.accounts()) {
        if (a.customerId == customerId) out.push_back(&a);
    }
    return out;
}

long long AccountService::applyMonthlyInterest() {
    long long total = 0;
    const std::string today = Utils::today();

    for (auto& acc : db_.accounts()) {
        if (acc.status != AccountStatus::ACTIVE) continue;
        if (acc.interestRate <= 0.0) continue;
        if (acc.lastInterestApplied == today) continue;  // once per day

        long double monthly = acc.interestRate / 100.0L / 12.0L;
        long double gain = static_cast<long double>(acc.balancePaise) * monthly;
        long long gainPaise = std::llround(gain);
        if (gainPaise <= 0) continue;

        acc.balancePaise += gainPaise;
        acc.lastInterestApplied = today;

        Transaction tx;
        tx.id = db_.nextTransactionId();
        tx.accountNumber = acc.accountNumber;
        tx.type = TxType::INTEREST;
        tx.amountPaise = gainPaise;
        tx.balanceAfterPaise = acc.balancePaise;
        tx.counterparty = "Bank";
        tx.timestamp = Utils::now();
        tx.remark = "Monthly interest @ " + std::to_string(acc.interestRate) + "% p.a.";
        db_.addTransaction(tx);

        total += gainPaise;
    }

    if (total > 0) db_.saveAll();
    return total;
}

bool AccountService::setInterestRate(const std::string& accountNumber,
                                     double rate, std::string& err) {
    Account* acc = db_.findAccountByNumber(accountNumber);
    if (!acc) {
        err = "Account not found.";
        return false;
    }
    if (acc->type != AccountType::SAVINGS) {
        err = "Only savings accounts earn interest.";
        return false;
    }
    if (rate < 0.0 || rate > 20.0) {
        err = "Rate must be between 0% and 20%.";
        return false;
    }
    acc->interestRate = rate;
    db_.saveAll();
    return true;
}

}  // namespace bms

File: web/app/services/account_service.py
"""Account management: open/close, interest rates and monthly interest accrual."""
from .. import money
from ..database import get_db, next_id
from . import notification_service

ACCOUNT_TYPES = ("Savings", "Checking")

MIN_OPENING_PAISE = 10000  # ₹100.00


def open_account(user_id, acct_type, opening_paise, remark="New account opened",
                 require_minimum=True):
    """Open a new account. Returns (account_number, None) or (None, error)."""
    if acct_type not in ACCOUNT_TYPES:
        return None, "Unknown account type."
    if require_minimum and opening_paise < MIN_OPENING_PAISE:
        return None, f"Minimum opening deposit is {money.format_paise(MIN_OPENING_PAISE)}."
    if opening_paise < 0:
        return None, "Opening deposit cannot be negative."

    db = get_db()
    number = f"OB{next_id(db, 'account_seq'):06d}"
    now = money.now()
    rate = 3.5 if acct_type == "Savings" else 0.0
    try:
        db.execute(
            "INSERT INTO accounts (account_number, user_id, type, balance_paise,"
            " interest_rate, status, created_at, last_interest_applied)"
            " VALUES (?, ?, ?, ?, ?, 'Active', ?, ?)",
            (number, user_id, acct_type, opening_paise, rate, now, money.today()),
        )
        db.execute(
            "INSERT INTO transactions (account_number, type, amount_paise,"
            " balance_after_paise, counterparty, timestamp, remark)"
            " VALUES (?, 'OPENING', ?, ?, ?, ?, ?)",
            (number, opening_paise, opening_paise, "OneSky Bank", now, remark),
        )
        notification_service.notify(
            user_id, "Account opened",
            f"A {acct_type} account {number} was opened. "
            f"Balance {money.format_paise(opening_paise)}.",
            "success",
        )
        db.commit()
    except Exception:
        db.rollback()
        return None, "Could not open the account."
    return number, None


def accounts_of(user_id):
    db = get_db()
    return db.execute(
        "SELECT * FROM accounts WHERE user_id = ? ORDER BY account_number",
        (user_id,),
    ).fetchall()


def get_account(number):
    return get_db().execute(
        "SELECT * FROM accounts WHERE account_number = ?", (number,)
    ).fetchone()


def set_rate(number, rate_raw):
    """Set the interest rate for a savings account."""
    try:
        rate = float(rate_raw)
    except (TypeError, ValueError):
        return "Interest rate must be a number."
    if rate < 0 or rate > 20:
        return "Interest rate must be between 0% and 20%."
    db = get_db()
    acct = get_account(number)
    if acct is None:
        return "Account not found."
    if acct["type"] != "Savings":
        return "Interest rate only applies to savings accounts."
    db.execute(
        "UPDATE accounts SET interest_rate = ? WHERE account_number = ?",
        (rate, number),
    )
    db.commit()
    return None


def close_account(number):
    """Close an account (empty balance required). Returns error or None."""
    db = get_db()
    acct = get_account(number)
    if acct is None:
        return "Account not found."
    if acct["status"] != "Active":
        return "Account is already closed."
    if acct["balance_paise"] != 0:
        return "Balance must be zero before the account can be closed."
    open_loans = db.execute(
        "SELECT 1 FROM loans WHERE account_number = ? AND status = 'Active'",
        (number,),
    ).fetchone()
    if open_loans:
        return "Clear outstanding loans before closing this account."
    db.execute(
        "UPDATE accounts SET status = 'Closed' WHERE account_number = ?", (number,)
    )
    db.execute(
        "INSERT INTO transactions (account_number, type, amount_paise,"
        " balance_after_paise, counterparty, timestamp, remark)"
        " VALUES (?, 'CLOSED', 0, 0, 'OneSky Bank', ?, 'Account closed')",
        (number, money.now()),
    )
    notification_service.notify(
        acct["user_id"], "Account closed",
        f"Account {number} was closed.",
        "info",
    )
    db.commit()
    return None


def apply_monthly_interest():
    """Apply one month of interest to every active savings account.

    Returns (count, total_paise) of interest credited.
    """
    db = get_db()
    rows = db.execute(
        "SELECT * FROM accounts WHERE status = 'Active' AND type = 'Savings'"
        " AND interest_rate > 0"
    ).fetchall()
    now = money.now()
    total = 0
    count = 0
    for acct in rows:
        interest = round(acct["balance_paise"] * acct["interest_rate"] / 100 / 12)
        if interest <= 0:
            continue
        new_balance = acct["balance_paise"] + interest
        db.execute(
            "UPDATE accounts SET balance_paise = ?, last_interest_applied = ?"
            " WHERE account_number = ?",
            (new_balance, money.today(), acct["account_number"]),
        )
        db.execute(
            "INSERT INTO transactions (account_number, type, amount_paise,"
            " balance_after_paise, counterparty, timestamp, remark)"
            " VALUES (?, 'INTEREST', ?, ?, 'OneSky Bank', ?, 'Monthly interest credit')",
            (acct["account_number"], interest, new_balance, now),
        )
        notification_service.notify(
            acct["user_id"], "Interest credited",
            f"{money.format_paise(interest)} interest was credited to {acct['account_number']} "
            f"at {acct['interest_rate']:g}% p.a. New balance {money.format_paise(new_balance)}.",
            "success",
        )
        total += interest
        count += 1
    db.commit()
    return count, total
